### PR TITLE
Added drag and drop function to the canvas

### PR DESCRIPTION
**Pull request by charlie0605**
Player are now able to drag avatar around on stage, when the player is dragging the avatar it will have a shadow effect of on avatar.
When the avatar is released it avatar will move to its new position.

When moving the avatar around drawing function will not work.

### CODE REVIEW

File: app/assets/javascripts/channels/theatre.coffee [hunk 1]
```diff
@@ -55,24 +56,40 @@ jQuery(document).on 'turbolinks:load', ->
...
     canvas.on 'mousedown touchstart', (e) ->
-      e.preventDefault()
-      x = e.offsetX
-      y = e.offsetY
-
-      if e.originalEvent.changedTouches
-        e = e.originalEvent.changedTouches[0]
+      if(window.holding != undefined)
+        avatar = App.state.avatars[window.holding]
+      if(avatar != undefined && e.clientX >= avatar.x && e.clientX <= (avatar.width + avatar.x) && e.clientY >= avatar.y && e.clientY <= (avatar.height + avatar.y))
```
Comment (rohrlach) [inline app/assets/javascripts/channels/theatre.coffee]: @charlie0605 could you please make this more readable?  Something like:
```
if avatar != undefined
   if e.clientX >= avatar.x && e.clientX <= (avatar.width + avatar.x)
      if e.clientY >= avatar.y && e.clientY <= (avatar.height + avatar.y)
         ...
```



I realize this makes nesting less than ideal - But it makes the code a LOT more readable IE:
- check avatar is defined
- check x 
- check y

File: app/assets/javascripts/channels/theatre.coffee [hunk 2]
```diff
@@ -55,24 +56,40 @@ jQuery(document).on 'turbolinks:load', ->
...
     canvas.on 'mousemove touchmove', (e) ->
-      if(drawing && $.now() - timeSinceLastSend > 10)
+      if(dragging)
+        avatar = App.state.avatars[window.holding]
+        img = new Image()
+        img.onload = ->
+          context.clearRect(0, 0, innerWidth, innerHeight);
+          App.drawFrame()
+          context.globalAlpha = 0.4
+          context.drawImage(img, e.clientX-avatar.width/2, e.clientY-avatar.height/2, avatar.width, avatar.height)
```
Comment (rohrlach) [inline app/assets/javascripts/channels/theatre.coffee]: Also here.

It's more often than not better to be clearer and keep the code clean than to worry about the performance hit of assigning an extra couple of local variables (Which, doesn't matter at all in our case) IE:
```
x = e.clientX - (avatar.width / 2)
y = e.clientY - (avatar.height / 2)
context.drawImage img, x, y, avatar.width, avatar.height
```

File: app/assets/javascripts/channels/theatre.coffee (hunk 1 above)
Comment (rohrlach) [reply app/assets/javascripts/channels/theatre.coffee]: In general, we should avoid nesting logic UNLESS that logic requires more than TWO expressions - IE: only one ```&&``` or one ```||```

Comment (charlie0605) [reply app/assets/javascripts/channels/theatre.coffee]: Ok ill keep that in mind

Comment (thongteav) [reply app/assets/javascripts/channels/theatre.coffee]: I see. I usually just do the box detection in one condition. Should change the habit of doing that now.

**Comment by rohrlach** [comment]
@charlie0605 I assume you've confirmed the changes still work as expected?

File: app/assets/javascripts/channels/theatre.coffee (hunk 1 above)
Comment (rohrlach) [reply app/assets/javascripts/channels/theatre.coffee]: Well remember, if you find yourself doing the same thing multiple times - You could always write a reusable function to do it 😉

**Comment by thongteav** [comment]
@rohrlach I've tested it on firefox, edge and chrome after the updated code. It is still working in the same behavior.

**Comment by helenvarley** [comment]
tested on firefox & chrome on linux, works. very awesome new feature, thanks!! :+1:

**Pull request merged**